napari/utils/colormaps/colormap.py

```diff
@@ -162,58 +162,134 @@ class LabelColormap(Colormap):
-        """Map *precast* values to colors.
+        Parameters
+        ----------
+        dtype : np.dtype
+            The dtype to convert the selection to.
 
-        When mapping values, we first convert them to a smaller dtype for
-        performance reasons. This conversion changes the label values,
-        even for small labels. This method is used to map values that have
-        already been converted to the smaller dtype.
+        Returns
+        -------
+        int
+            The selection converted.
+        """
+        return _cast_labels_to_minimum_dtype_auto(
+            np.array([self.selection]).astype(dtype), self
+        )[0]
+
+    def background_as_minimum_dtype(self, dtype: np.dtype) -> int:
+        """Treat selection as given dtype and calculate its
+        value to minimum dtype using _cast_labels_to_minimum_dtype_auto
+        function.
 
         Parameters
         ----------
-        values : np.ndarray
-            Values to be mapped. They must have already been downcast using
-            `_cast_labels_to_minimum_dtype_auto`.
+        dtype : np.dtype
+            The dtype to convert the background to.
 
         Returns
         -------
-        np.ndarray of shape (N, M, 4)
+        int
+            The background converted.
+        """
+        return _cast_labels_to_minimum_dtype_auto(
+            np.array([self.background_value]).astype(dtype), self
+        )[0]
+
+    def map(self, values, apply_selection=True) -> np.ndarray:
+        """Map values to colors.
+
+        Parameters
+        ----------
+        values : np.ndarray or float
+            Values to be mapped.
+        apply_selection : bool
+            Whether to apply selection if self.use_selection is True.
+
+        Returns
+        -------
+        np.ndarray of same shape as values, but with last dimension of size 4
             Mapped colors.
         """
-        mapped = self.colors[values.astype(np.int64)]
+        values = np.atleast_1d(values)
 
-        mapped[values == self.background_value] = 0
+        if values.dtype.kind == 'f':
+            values = values.astype(np.int64)
 
-        # If using selected, disable all others
-        if self.use_selection:
-            cast_selection = _cast_labels_to_minimum_dtype_auto(
-                np.array([self.selection]),
-                len(self.colors) - 1,
-                self.background_value,
-            )[0]
-            mapped[values != cast_selection] = 0
+        if values.dtype == np.uint8 and "_uint8_colors" in self.__dict__:
+            mapped = self._uint8_colors[values]
+        elif values.dtype == np.uint16 and "_uint16_colors" in self.__dict__:
+            mapped = self._uint16_colors[values]
+        else:
+            background = self.background_as_type(values.dtype)
+            # cast background to values dtype is to support int8 and int16 negative backgrounds
+            texture_dtype_values = _zero_preserving_modulo_naive(
```

Comment:
Can you rename this to `_zero_preserving_modulo_numpy` rather than naive? Naive is not the right word for this.

Reply:
done